File: python/0904_fruit_into_baskets.py
from collections import defaultdict

class Solution:
    def totalFruit(self, tree: List[int]) -> int:
        tlen = len(tree)
        if tlen == 0:
            return 0

        left = right = 0
        res = curSum = 0
        fruitMap = defaultdict(lambda: 0)

        shouldBreak = False
        while not shouldBreak:
            while len(fruitMap.keys()) <= 2:
                fruitMap[tree[right]] += 1
                curSum += 1
                if len(fruitMap.keys()) <= 2:
                    res = max(res, curSum)
                right += 1

                if right == tlen:
                    shouldBreak = True
                    break

            while left < right and len(fruitMap.keys()) > 2:
                curSum -= 1
                fruitMap[tree[left]] -= 1
                if fruitMap[tree[left]] == 0:
                    del fruitMap[tree[left]]
                left += 1

        return res

File: go/0904_fruit_into_baskets.go
func totalFruit(tree []int) int {
	res := 0
	m := make(map[int]int)

	for left, right := 0, 0; right < len(tree); right++ {
		if cnt, ok := m[tree[right]]; ok {
			m[tree[right]] = cnt + 1
		} else {
			m[tree[right]] = 1
		}

		for len(m) > 2 {
			if cnt, ok := m[tree[left]]; ok {
				m[tree[left]] = cnt - 1
				if cnt == 1 {
					delete(m, tree[left])
				}
			}
			left++
		}

		if dist := right - left + 1; dist > res {
			res = dist
		}
	}
	return res
}